### PR TITLE
Add intrinsics for lzcnt, tzcnt, bsr (amd64)

### PR DESCRIPTION
**Pull request by gretay-js**
Only the last commit.  

Support for lzcnt and tzcnt is off by default, but can be turned on using compile time flags -flznct -ftzcnt.
Warning: these instructions will silently produce wrong results when the code runs on CPUs prior to Haswell (2008).

### CODE REVIEW

**Comment by Kakadu** [comment]
Could you, please, recommend some  reading about the cases when these new instructions are especially great, and in particular when they are going to be especially useful in context of OCaml?

**Comment by gretay-js** [comment]
> Could you, please, recommend some reading about the cases when these new instructions are especially great, and in particular when they are going to be especially useful in context of OCaml?

These operations are used in some algorithms for fast radix trees, histogram and arbitrary precision arithmetic. When implementing such algorithms in OCaml, this PR provides a way to directly use the hardware instruction, when available. However, the concept is not special for OCaml. GCC and other compilers also provide these builtins for use in C/C++ code.

**Comment by gretay-js** [comment]
I've rebased it and converted to draft because I need to update the symbol names to those match ocaml_intrisics library and test it again.

File: backend/amd64/emit.mlp [hunk 1]
```diff
@@ -860,6 +860,20 @@ let emit_instr fallthrough i =
...
       I.mov (arg32 i 0) (res32 i 0)
+  | Lop(Ispecific(Ibsr { arg_is_non_zero; })) ->
+    if arg_is_non_zero then begin
+      (* No need to handle that bsr is undefined on 0 input. *)
+      I.bsr (arg i 0) (res i 0);
+    end else begin
+      let lbl_z = new_label () in
+      let lbl_nz = new_label () in
+      I.bsr (arg i 0) (res i 0);
+      I.je (label lbl_z);
```
Comment (smuenzel) [inline backend/amd64/emit.mlp:871]: If we're okay destroying the argument (or using a temporary)
```
I.bsr (arg i 0) (res i 0)
I.mov (int (-1)) (arg i 0)
I.cmov E (arg i 0) (res i 0)
```

Comment (gretay-js) [reply backend/amd64/emit.mlp:871]: Nice trick using `cmov` from #172 and `mov` doesn't set the flags.  We can do it for clz and ctz with non-zero args too.

Comment (lthls) [reply backend/amd64/emit.mlp:871]: Isn't there a problem if the argument and the result are the same register ?

Comment (gretay-js) [reply backend/amd64/emit.mlp:871]: yes, it would need to use another register.

**Comment by gretay-js** [comment]
I'm closing this and will submit a new PR to implement Cclz/Cctz operations using lzcnt/tzcnt instruction (instead of bsr/bsf) when depending on compiler configuration.

**Pull request closed without merging**